### DISCUSSION TITLE
Ability to specify `*.gem` path in `Gemfile`

### DISCUSSION THREAD

**Discussion by DannyBen** [ruby/rubygems#6096] [Q&A]
Hey all,

I was trying to figure out if this is possible, but [this page](https://bundler.io/man/gemfile.5.html) did not reveal the answer.

I have an internal, private gem, that I wish to avoid pushing anywhere (not even a private gem repo).
I wish to just put it in directory in the project, and have it install normally when I run `bundle install`

I was hoping for something like this:

```ruby
# Gemfile
source "https://rubygems.org"
gem "mygem", gem: 'mygem-0.0.1.gem'
```

or even better, without specifying the exact filename (which includes version), to just point bundler to a directory containing a file named `<gem_name>-*.gem`:

```ruby
# Gemfile
source "https://rubygems.org"
gem "mygem", gempath: 'mygem-dir'
```

and in `mygem-dir` have `mygem-0.0.1.gem`, and bundler will search for/install the latest version or the one specified in the lock file.

Is something like this possible? If not, any proposed alternative for this scenario (other than the usual install from source) are welcome.

**Comment by simi** [comment]
You can point to gem source code using `gem 'abc', path: 'vendor/abc'`. You can vendor your gem in git repository for example or alternatively you can extract your gem somewhere. `gem` file on its own is just `tar`.

In theory you should be able to do:

```bash
mkdir /gems/mygem
mv mygem-0.0.1.gem /gems/mygem
cd /gems/mygem
tar -xvf mygem-0.0.1.gem # this extracts 3 file -> metadata.gz, data.tar.gz, checksums.yaml.gz
tar -xvzf data.tar.gz # extract data
```

and in Gemfile point to that path using `gem 'mygem', path: '/gems/mygem'`.

**Reply by DannyBen** [reply]
Yeah, I am familiar with `path`, and I was hoping it can accept path to a `*.gem` file.
It surprised me that out of all the options for sourcing a gem, providing path to a .gem file is not available.

**Reply by simi** [reply]
AFAIK there is no way to source gem file directly. But it is interesting idea, I think that PR would be accepted.

**Reply by simi** [reply]
It could work in similar way to pointing to git, just install the gem from local source into bundle path.

**Reply by DannyBen** [reply]
> But it is interesting idea, I think that PR would be accepted.

I wish I could. This seems above my paygrade.

> It could work in similar way to pointing to git, just install the gem from local source into bundle path.

I imagined it more like the `path` argument, even using the same attribute.

**Comment by DannyBen** [comment]
By the way, these instructions

```bash
tar -xvf mygem-0.0.1.gem
tar -xvzf data.tar.gz
```

```ruby
# Gemfile
gem 'mygem', path: '/gems/mygem'
```

do not work for me unless the directory also contains the gemspec file...

**Reply by simi** [reply]
Ahh, you're right. That should be part of metadata, but serialized probably.

**Reply by DannyBen** [reply]
Yup. Ended up trying with a simple zip file containing `lib/*` and the gemspec, which worked of course, but felt really hacky.
Was quite sure this would be a built in feature. I will abandon it for now.

Thanks for the help.

**Comment by mensfeld** [comment]
@simi what would be the expected option? should it work with `path`, `file` or something else?
